// generators/srl.js

/**
 * 语义角色标注（srl）问题
 */

var pushQuestions = require("../tools/pushQuestions.js");
var componentReplace = require("../tools/componentReplace.js")

module.exports = srlGenerateQuestionsBySentence;

function srlGenerateQuestionsBySentence(sentence) {
    var questions = [];
    for(var wordIndex = 0; wordIndex<sentence.length; wordIndex++){
        var word = sentence[wordIndex];
        var newQuestions = srlGenrateQuestionsByWord(sentence,word);
        for(var qi = 0;qi<newQuestions.length;qi++){
            questions = pushQuestions(questions,newQuestions[qi]);
        }
    }
    return questions;
}

function srlGenrateQuestionsByWord(sentence,word) {
    var questions = [];
    if(word.arg.length != 0) {
        //先查找是否有component.type=='A0'||'A1'的词，如果有，将这个词赋值给A0|A1
        for(var componentIndex = 0; componentIndex<word.arg.length; componentIndex++){
            var component = word.arg[componentIndex];
            if (component.type=='A0') {
                A0 = component;
            };
            if (component.type=='A1') {
                A1 = component;
            };
        }
        for(var componentIndex = 0; componentIndex<word.arg.length; componentIndex++){
            var component = word.arg[componentIndex];
            var newQuestion = srlGenrateQuestionByComponent(sentence, component);
            if(newQuestion)
                questions = pushQuestions(questions, newQuestion);
        }
    }
    return questions;
}

//component是当前词的arg
function srlGenrateQuestionByComponent(sentence, component) {
    var ruleFactory = function(srl,label,rpc){
        return {
            srl : srl,
            label : label,
            replacement : rpc
        };
    };
    var rules = [
        ruleFactory('LOC','地点状语','在哪里'),
        ruleFactory('TMP','时间状语','什么时候'),
        ruleFactory('A0','A0','谁')
    ];
    for(var i = 0; i<rules.length; i++) {
        var rule = rules[i];
        if(component.type == rule.srl) {
            var replacement = rule.replacement;
            var text = null;
            if(component.type == 'TMP') {
                replacement = getTMPReplacement(sentence,component);              
            }
            if(component.type == 'A0') {
                A0 = component;
                var currA0pos = sentence[component.end].pos;
                if(/(n|nd|nz|ni)$/.test(currA0pos)) {
                    replacement = '什么';
                    //console.log(currA0pos);
                    text = componentReplace(sentence,component,replacement);
                }else {
                    text = componentReplace(sentence,component,replacement);
                }  
            }
            else{
                // if(A0){
                //     var sentence_ = removeComponent(sentence,component);
                //     replacement = getComponentStr(sentence,A0) + replacement;
                //     console.log(replacement);
                //     text = componentReplace(sentence_,A0,replacement);
                // }
                // else{
                    text = componentReplace(sentence,component,replacement);
                // }
            }
            var question = {
                label : rule.label, 
                text : text
            };
            return question;
        }
    }
}



//得到component对应的词
function getComponentStr(sentence,component) {
    var str = ""
    for (var i = component.beg; i<=component.end; i++) {
        str += sentence[i].cont;
    }
    return str;
}

//得到对时间提问的疑问词
function getTMPReplacement(sentence, component) {
    var componentStr = getComponentStr(sentence, component);
    if(componentStr.indexOf('日')>0) {
        return '哪一天';
    }
    if(componentStr.indexOf('月')>0) {
        return '几月份';
    }
    if(componentStr.indexOf('年')>0){
        return '哪一年';
    }
    if(componentStr.indexOf('世纪')>0) {
        return '几世纪';
    }
    else
        return '什么时候';
}

//得到去掉component后的句子
function removeComponent(sentence,component) {
    var sent = [];
    for(var i = 0; i<sentence.length; i++) {
        var word = sentence[i];
        if(i<component.beg || i>component.end) {
            sent.push(word);
        }
    }
    return sent;
}
